mlrun/launcher/local.py

```diff
@@ -23,17 +40,291 @@ def verify_base_image(runtime):
+
+        if "V3IO_USERNAME" in os.environ and "v3io_user" not in run.metadata.labels:
+            run.metadata.labels["v3io_user"] = os.environ.get("V3IO_USERNAME")
+
+        logger.info(
+            "Storing function",
+            name=run.metadata.name,
+            uid=run.metadata.uid,
+            db=runtime.spec.rundb,
+        )
+        self.store_function(runtime, run, run.metadata)
+
+        from mlrun.run import MLClientCtx
+
+        execution = MLClientCtx.from_dict(
+            run.to_dict(),
+            self.db,
+            autocommit=False,
+            is_api=False,
+            store_run=False,
+        )
+
+        from mlrun.runtimes.generators import get_generator
+
+        # create task generator (for child runs) from spec
+        task_generator = get_generator(run.spec, execution)
+        if task_generator:
+            # verify valid task parameters
+            tasks = task_generator.generate(run)
+            for task in tasks:
+                runtime._validate_run_params(task.spec.parameters)
+
+        # post verifications, store execution in db and run pre run hooks
+        execution.store_run()
+        self._pre_run(runtime, run, execution)  # hook for runtime specific prep
+
+        last_err = None
+        # If the runtime is nested, it means the hyper-run will run within a single instance of the run.
+        # So while in the API, we consider the hyper-run as a single run, and then in the runtime itself when the
+        # runtime is now a local runtime and therefore `self._is_nested == False`, we run each task as a separate run by
+        # using the task generator
+        # TODO client-server separation might not need the not runtime._is_nested anymore as this executed local func
+        if task_generator and not runtime._is_nested:
+            # multiple runs (based on hyper params or params file)
+            runner = runtime._run_many
+            if hasattr(runtime, "_parallel_run_many") and task_generator.use_parallel():
+                runner = runtime._parallel_run_many
+            results = runner(task_generator, execution, run)
+            mlrun.runtimes.utils.results_to_iter(results, run, execution)
+            result = execution.to_dict()
+            result = runtime._update_run_state(result, task=run)
+
+        else:
+            # single run
+            try:
+                resp = runtime._run(run, execution)
+                result = runtime._update_run_state(resp, task=run)
+            except mlrun.runtimes.base.RunError as err:
+                last_err = err
+                result = runtime._update_run_state(task=run, err=err)
+
+        self._save_or_push_notifications(run)
+        # run post run hooks
+        self._post_run(execution)  # hook for runtime specific cleanup
+
+        return runtime._wrap_run_result(result, run, err=last_err)
+
+    @staticmethod
+    def _pre_run(runtime, runobj, execution):
```

Comment:
pre run is per runtime (AbstractSparkRuntime also implements it) and this is used in LocalRuntime in 2 contexts:
1. the user passed local=True when initiating the run (hence the _is_run_local below)
2. inside a pod that was initiated by the API

I think we still need the _is_run_local but we should make it normal bool instead of using setattr and hasattr